### ISSUE TITLE
feat: add localLink that runs the router in the same process as the client

### ISSUE THREAD

**Issue by dbartholomae** [trpc/trpc#3768]
### Describe the feature you'd like to request

Hi there!

It would be great to have a `LocalLink` that allows me to do:
This would directly run the router in the same process as the client instead of transferring requests to a server.

### But why??

This is motivated mainly due to testing and fast prototyping:
1. This allows to test the client with the server code (with a small caveat, see below)
2. I can start writing a fully functional client that saves data in local storage instead of a database, and when I want to add a database, I don't need to rewrite the procedures, giving me significantly faster prototyping speed.

### Describe the solution you'd like to see

```ts
import { createRouter } from '../path/to/server/trpc';
const client = createTRPCProxyClient<ReturnType<typeof createRouter>>({
  links: [
    localLink({ router: createRouter({ database }) }),
  ],
});```

For the feature itself it doesn't matter that the router is created by a function call, but this solves a problem I mentioned above: What if part of the code only runs on the server, e.g. database access? Well, just inject that dependency, and, in case you are running on the client, use a browser-based database.

### Describe alternate solutions

Not sure, to be honest.

### Additional information

_No response_

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

**Comment by KATT** [comment]
I've seen people doing this and call it ssrLink

**Comment by dbartholomae** [comment]
Good point, this also makes sense for Server-Side-Rendering. Can you point me to an example? And would you be interested inhaving a link like this as part of trpc core?

**Comment by KATT** [comment]
@ottomated made this https://github.com/ottomated/trpc-svelte-query/blob/main/src/ssr/ssrLink.ts

**Comment by dbartholomae** [comment]
Thanks! The SSR link is still using HTTP to communicate with a server instead of running the router itself. I was more thinking about something like:

```ts
export function localLink<TRouter extends AnyRouter>({
  router,
}: {
  router: TRouter;
}): TRPCLink<TRouter> {
  return () =>
    ({ op }) =>
      observable<OperationResultEnvelope<unknown>, TRPCClientError<TRouter>>(
        (observer) => {
          async function execute() {
            const caller = router.createCaller({});
            try {
              const data = await (
                caller[op.path] as (input: unknown) => unknown
              )(op.input);
              observer.next({ result: { data, type: "data" } });
              observer.complete();
            } catch (err) {
              observer.error(TRPCClientError.from(err as Error));
            }
          }

          void execute();
        }
      );
}
```

One problem I ran into is that I get an error message when running the router in the browser. Is there a reason why the router has to be run on a server? I would have assumed that only the adapters (and maybe some self-written code) would need to worry about whether the code runs on server or on client.

**Comment by KATT** [comment]
You can use `initTRPC.create({ allowOutsideOfServer: true })`

In most cases, it's a mistake to have it running in the browser, you likely don't intend to import your server code on the client. 

Even in this case, you probably don't want to have this running in the browser, but maybe in your tests/prototyping you can flag that as `true`

**Comment by dbartholomae** [comment]
Thanks, that's helpful! The reason why I sometimes have router running on the client is for rapid prototyping: This way I can build an app that saves data in localStorage instead of saving it on the server, and then move the router to the backend once I actually need communication across clients.

Coming back to my question for `localLink`: Would you be interested in adding this into trpc? Then I would open up a PR along the lines of what I posted before, but a bit cleaned, with test coverage, and some additions (e.g. the ability to set context) and with some documentation added how to use this for testing. Otherwise, I would most likely just set this up as a separate repo and open a PR to the documentation "Awesome tRPC Collection" page.
Both is fine for me, I would have a slight preference to add it to this repo, though, as it adds a nice way to write tests for the frontend when using tRPC, which so far is a bit missing as there is no easy way to correctly mock the network calls made by the adapters.

**Comment by dtinth** [comment]
I have another use case for having local links in tRPC — allowing a server-side rendered web app to choose whether to use a remote backend or a local backend. See below picture for explanation of the scenario.

<img width="1153" alt="image" src="https://user-images.githubusercontent.com/193136/235290100-47a18440-0a17-4acf-8218-69524cdfe6c3.png">

The open source website we’re building has a **backend logic** (which requires a lot of secret credentials) and a **server-side rendered web app.** They are on the same project and deployed as a single unit, but to keep things modular and composable (and allow 3rd party to integrate), the server-side web app delegates all backend logic to a tRPC router.

Currently we work around this by hijacking the HTTP link to intercept tRPC requests to `http://fake` and throw it to the tRPC Server’s `fetchRequestHandler`. This way the router code is executed on the same process without any involvement of network.

```ts
import { fetchRequestHandler } from '@trpc/server/adapters/fetch'
import { createTRPCProxyClient, httpLink } from '@trpc/client';
import type { AppRouter } from '../server';

export function createServerSideClient(authToken?: string) {
  const url = process.env.BACKEND || 'http://fake'

  return createTRPCProxyClient<AppRouter>({
    links: [
      httpBatchLink({
        url,
        headers: authToken ? {
          authorization: `Bearer ${authToken}`
        } : {},
        fetch: async (input, init) => {
          if (
            typeof input === 'string' &&
            input.startsWith('http://fake')
          ) {
            const request = new Request(input, init as RequestInit)
            const { createContextFromRequest, AppRouter } = await import('../server')
            return fetchRequestHandler({
              endpoint: prefix,
              req: request,
              router: AppRouter,
              createContext: ({ req }) => {
                return createContextFromRequest(req)
              },
            })
          }
          return fetch(input, init as RequestInit)
        },
      }),
    ],
  })
}
```

(Moved from this issue:)

- #4275

**Comment by dbartholomae** [comment]
[This](https://github.com/trpc/trpc/issues/3768#issuecomment-1431247072) code worked well for me. I'm still open to contribute this in bit more stable and tested way to this repo - but might just create a separate package if that is easier :)

**Comment by dtinth** [comment]
After seeing this issue I improved my implementation to make it cleaner. Here’s my implementation of `localLink`, based on HTTP router (I do it this way because I don’t want to deal with Observables). It also supports passing a context to the procedure:

```ts
import { httpLink } from '@trpc/client'
import type { AnyRouter, inferRouterContext } from '@trpc/server'
import { fetchRequestHandler } from '@trpc/server/adapters/fetch'

export function localLink<TRouter extends AnyRouter>(
  router: TRouter,
  context: inferRouterContext<TRouter>
) {
  return httpLink({
    url: 'http://local',
    fetch: async (...args) => {
      const request = new Request(...args)
      return fetchRequestHandler({
        endpoint: '',
        req: request,
        router,
        createContext: () => context,
      })
    },
  })
}
```

**Comment by KATT** [comment]
This is close to what you want: https://github.com/trpc/trpc/blob/9d5561d1b6b74c7ac9df393f6a3344cf7d93b7f2/packages/next/src/app-dir/create-action-hook.tsx#L71-L100

**Comment by dbartholomae** [comment]
Not sure I understand how this would help

**Comment by ssendev** [comment]
> This is close to what you want:

And how to use it?

I managed to modify it like this:
```ts
export function experimental_serverActionLink<
  TRouter extends AnyRouter = AnyRouter
>(router: any): TRPCLink<TRouter> {
  return (runtime) =>
    ({ op }) =>
      observable((observer) => {
        const path = op.path.split(':')
        let r = router
        for (const p of path) {
          r = r[p]
        }

        r(
          isFormData(op.input)
            ? op.input
            : runtime.transformer.serialize(op.input)
        )
          .then((data: any) => {
            const transformed = transformResult({ result: { data } }, runtime)

            if (!transformed.ok) {
              observer.error(TRPCClientError.from(transformed.error, {}))
              return
            }
            observer.next({
              context: op.context,
              result: transformed.result,
            })
            observer.complete()
          })
          .catch((cause: any) => observer.error(TRPCClientError.from(cause)))
      })
}
```
and could then use it like this:
```ts
let trpc
if (import.meta.env.SSR) {
  trpc = createTRPCProxyClient<AppRouter>({
    links: [
      experimental_serverActionLink(
        appRouter.createCaller(
          createMyContext({ req: context.req, res: context.res })
        )
      ),
    ],
  })
} else {
  trpc = createTRPCProxyClient<AppRouter>({
    links: [httpBatchLink({  '/api'  })],
  })
}
await trpc.my.own.query( 'arg')
```

My use case is SSR and don't want to special case the RPCs so need the same API on server and client.

**Comment by nwalters512** [comment]
I'd love to see this built in to tRPC! In the meantime, the suggestion in https://github.com/trpc/trpc/issues/3768#issuecomment-1528829502 worked great for me.

A tiny bit of context on my use case: in local development, I run all code in a single process for simplicity, but in production, I deploy portions of my app on dedicated hardware. I'd like to abstract things so that callsites don't need to know or care about whether something is happening across a network boundary or not. A "local" link seems to be the perfect way to achieve that.
